client/dashboard/domains/domain-dns/email-setup-form.tsx

```diff
@@ -0,0 +1,147 @@
+import { useMutation } from '@tanstack/react-query';
+import {
+	__experimentalHStack as HStack,
+	__experimentalVStack as VStack,
+	Button,
+} from '@wordpress/components';
+import { useDispatch } from '@wordpress/data';
+import { DataForm, Field, isItemValid } from '@wordpress/dataviews';
+import { __, sprintf } from '@wordpress/i18n';
+import { store as noticesStore } from '@wordpress/notices';
+import { useState } from 'react';
+import { domainDnsApplyTemplateMutation } from '../../app/queries/domain-dns-records';
+import { domainRoute } from '../../app/routes/domain-routes';
+
+export type EmailSetupFormData = {
+	record: string;
+};
+
+const form = {
+	type: 'regular' as const,
+	fields: [ 'record' ],
+};
+
+const defaultFormData: EmailSetupFormData = {
+	record: '',
+};
+
+interface EmailSetupFormProps {
+	description: string;
+	label: string;
+	modifyVariables?: ( variables: { token: string; domain: string; mxdata?: string } ) => {
+		token: string;
+		domain: string;
+		mxdata?: string;
+	};
+	pattern: RegExp;
+	placeholder: string;
+	provider: string;
+	service: string;
+}
+
+export default function EmailSetupForm( {
+	description,
+	label,
+	modifyVariables,
+	pattern,
+	placeholder,
+	provider,
+	service,
+}: EmailSetupFormProps ) {
+	const { domainName } = domainRoute.useParams();
+	const [ formData, setFormData ] = useState< EmailSetupFormData >( defaultFormData );
+	const [ isPending, setIsPending ] = useState( false );
+	const { createSuccessNotice, createErrorNotice } = useDispatch( noticesStore );
+
+	const mutation = useMutation( domainDnsApplyTemplateMutation( domainName ) );
+
+	const handleSubmit = ( e: React.FormEvent ) => {
+		e.preventDefault();
+		setIsPending( true );
+		let variables: {
+			token: string;
+			domain: string;
+			mxdata?: string;
+		} = {
+			token: formData.record,
+			domain: domainName,
+		};
+		if ( modifyVariables ) {
+			variables = modifyVariables( variables );
+		}
+		mutation.mutate(
+			{
+				provider,
+				service,
+				variables,
+			},
+			{
+				onSuccess: () => {
+					createSuccessNotice( __( 'Email setup completed successfully.' ), {
+						type: 'snackbar',
+					} );
+				},
+				onError: () => {
+					createErrorNotice( __( 'Failed to complete email setup.' ), {
+						type: 'snackbar',
+					} );
+				},
+				onSettled: () => {
+					setIsPending( false );
+					setFormData( defaultFormData );
+				},
+			}
+		);
+	};
+
+	const fields: Field< EmailSetupFormData >[] = [
+		{
+			id: 'record',
+			type: 'text',
+			label: __( 'Verification token' ),
+			description,
+			placeholder,
+			isValid: {
+				custom: ( item ) => {
+					return pattern.test( item.record ) ? null : __( 'Invalid verification record format' );
+				},
+				required: true,
+			},
+		},
+	];
+
+	const canSubmit = isItemValid( formData, fields, form );
```

Comment:
I did not know about this method. I'm going to implement on my Namer Servers form. Thanks for that! 😄